esmvaltool/diag_scripts/weighting/weighted_temperature_map.py

```diff
@@ -0,0 +1,285 @@
+    """
+    values = np.array(values)
+    if weights is None:
+        weights = np.ones(len(values))
+    weights = np.array(weights)
+
+    if not ((quantile >= 0) & (quantile <= 1)):
+        raise ValueError('Quantile should be between 0.0 and 1.0')
+
+    idx = np.argsort(values)
+    values = values[idx]
+    weights = weights[idx]
+
+    weighted_quantiles = np.cumsum(weights) - 0.5 * weights
+
+    # Cast weighted quantiles to 0-1 To be consistent with np.quantile
+    min_val = weighted_quantiles.min()
+    max_val = weighted_quantiles.max()
+    weighted_quantiles = (weighted_quantiles - min_val) / max_val
+
+    return np.interp(quantile, weighted_quantiles, values)
+
+
+def visualize_and_save_temperature(temperature: 'xr.DataArray',
+                                   cfg: dict,
+                                   ancestors: list):
+    """Visualize weighted temperature."""
+    period = '{start_year}-{end_year}'.format(**read_metadata(cfg)['tas'][0])
+    if (meta := read_metadata(cfg).get('tas_reference', None)) is not None:
+        period = '{} minus {start_year}-{end_year}'.format(period, **meta[0])
+    metric = cfg['model_aggregation']
+    if isinstance(metric, int):
+        metric = f'{metric}perc'
+
+    proj = ccrs.PlateCarree(central_longitude=0)
+    fig, ax = plt.subplots(subplot_kw={'projection': proj})
+
+    temperature.plot.pcolormesh(
+        ax=ax,
+        transform=ccrs.PlateCarree(),
+        levels=9,
+        robust=True,
+        extend='both',
+        cmap='Reds',
+    )
+
+    ax.coastlines()
+    lons = temperature.lon.values
+    lats = temperature.lat.values
+    longitude_formatter = LongitudeFormatter()
+    latitude_formatter = LatitudeFormatter()
+    if (xticks := cfg.get('xticks', None)) is not None:
+        ax.set_xticks(xticks)
+    else:
+        ax.set_xticks(np.arange(np.floor(lons.min()), np.ceil(lons.max()), 10), crs=proj)
+    if (yticks := cfg.get('yticks', None)) is not None:
+        ax.set_yticks(yticks)
+    else:
+        ax.set_yticks(np.arange(np.floor(lats.min()), np.ceil(lats.max()), 10), crs=proj)
+    ax.xaxis.set_ticks_position('both')
+    ax.yaxis.set_ticks_position('both')
+    ax.xaxis.set_major_formatter(longitude_formatter)
+    ax.yaxis.set_major_formatter(latitude_formatter)
+    ax.set_xlabel('')
+    ax.set_ylabel('')
+
+    ax.set_title(f'Weighted {metric} temperature \n{period} ($\degree$C)')
+
+    filename_plot = get_plot_filename('temperature_change_weighted_map', cfg)
+    fig.savefig(filename_plot, dpi=300, bbox_inches='tight')
+    plt.close(fig)
+
+    filename_data = get_diagnostic_filename(
+        'temperature_map_weighted', cfg, extension='nc')
+    temperature.to_netcdf(filename_data)
+
+    caption = f'Temperature map {period}'
+    log_provenance(caption, filename_plot, cfg, ancestors)
+    log_provenance(caption, filename_data, cfg, ancestors)
+
+
+def visualize_and_save_temperature_difference(temperature: 'xr.DataArray',
```

Comment:
The bulk of this function looks very similar to the previous one. Would it be possible to make one slightly more generic function that plots a map of temperature given it an array, and then just pass 2 different arrays to it? One with the mean, and one with the difference?

Reply:
Merged them!